lib/ConsumerImplBase.cc

```diff
@@ -0,0 +1,145 @@
+    }
+    batchReceiveTimer_ = listenerExecutor_->createDeadlineTimer();
+}
+
+void ConsumerImplBase::triggerBatchReceiveTimerTask(long timeoutMs) {
+    if (timeoutMs > 0) {
+        batchReceiveTimer_->expires_from_now(boost::posix_time::milliseconds(timeoutMs));
+        std::weak_ptr<ConsumerImplBase> weakSelf{shared_from_this()};
+        batchReceiveTimer_->async_wait([weakSelf](const boost::system::error_code& ec) {
+            // If two requests call runPartitionUpdateTask at the same time, the timer will fail, and it
+            // cannot continue at this time, and the request needs to be ignored.
+            auto self = weakSelf.lock();
+            if (self && !ec) {
+                self->doBatchReceiveTimeTask();
+            }
+        });
+    }
+}
+
+void ConsumerImplBase::doBatchReceiveTimeTask() {
+    if (state_ != Ready) {
+        return;
+    }
+
+    bool hasPendingReceives = false;
+    long timeToWaitMs = batchReceivePolicy_.getTimeoutMs();
+
+    Lock lock(batchPendingReceiveMutex_);
+    while (!batchPendingReceives_.empty()) {
+        OpBatchReceive& batchReceive = batchPendingReceives_.front();
+        long diff =
+            batchReceivePolicy_.getTimeoutMs() - (TimeUtils::currentTimeMillis() - batchReceive.createAt_);
+        if (diff <= 0) {
+            notifyBatchPendingReceivedCallback(batchReceive.batchReceiveCallback_);
+            batchPendingReceives_.pop();
+        } else {
+            hasPendingReceives = true;
+            timeToWaitMs = diff;
+            break;
+        }
+    }
+    lock.unlock();
+
+    if (hasPendingReceives) {
+        triggerBatchReceiveTimerTask(timeToWaitMs);
+    }
+}
+
+void ConsumerImplBase::failPendingBatchReceiveCallback() {
+    Messages msgs;
+    Lock lock(batchPendingReceiveMutex_);
+    while (!batchPendingReceives_.empty()) {
+        OpBatchReceive opBatchReceive = batchPendingReceives_.front();
+        batchPendingReceives_.pop();
+        auto self = shared_from_this();
+        listenerExecutor_->postWork([opBatchReceive, self, msgs]() {
+            opBatchReceive.batchReceiveCallback_(ResultAlreadyClosed, msgs);
+        });
+    }
+    lock.unlock();
+}
+
+void ConsumerImplBase::notifyBatchPendingReceivedCallback() {
+    Lock lock(batchPendingReceiveMutex_);
+    if (!batchPendingReceives_.empty()) {
+        OpBatchReceive& batchReceive = batchPendingReceives_.front();
+        batchPendingReceives_.pop();
+        notifyBatchPendingReceivedCallback(batchReceive.batchReceiveCallback_);
+    }
+}
+
+void ConsumerImplBase::batchReceiveAsync(BatchReceiveCallback callback) {
+    // fail the callback if consumer is closing or closed
+    if (state_ != Ready) {
+        callback(ResultAlreadyClosed, Messages());
+        return;
+    }
+
+    if (hasEnoughMessagesForBatchReceive()) {
+        Lock lock(batchPendingReceiveMutex_);
+        notifyBatchPendingReceivedCallback(callback);
+        lock.unlock();
```

Comment:
Yes, you are right.  `batchPendingReceiveMutex_` is only used to make the accesses to `batchPendingReceives_` thread safe. I mixed it up.

Here need a lock to make `hasEnoughMessagesForBatchReceive` and `notifyBatchPendingReceivedCallback` is atomic, then prevents users receive not compliant with the policy messages.

I add a new `batchReceiveOptionMutex_` lock to sync them. 

But, This is not entirely guaranteed, Because I didn't synchronize the scenario where a single receive message is `incomingQueue_` out of the queue.

In Java client impl, These operations are all performed in `internalPinnedExecutor` threads. So no problem.

I think it can be done this way first, and then I'll try to refactor it so that C++ also uses an internal thread to avoid adding various locks.